File: snrg_credit_control/snrg_credit_control/report/credit_control_report/credit_control_report.js
// Credit Control Report — filters + formatting

frappe.query_reports["Credit Control Report"] = {
  filters: [
    {
      fieldname: "company",
      label: __("Company"),
      fieldtype: "Link",
      options: "Company",
      default: frappe.defaults.get_user_default("Company"),
      reqd: 1,
    },
    {
      fieldname: "status",
      label: __("Status"),
      fieldtype: "Select",
      options: [
        "All",
        "Credit Hold",
        "Pending Approval",
        "Approved",
        "Expired",
      ].join("\n"),
      default: "All",
    },
    {
      fieldname: "customer",
      label: __("Customer"),
      fieldtype: "Link",
      options: "Customer",
      get_query: () => ({ filters: { disabled: 0 } }),
    },
    {
      fieldname: "from_date",
      label: __("SO Date From"),
      fieldtype: "Date",
      default: frappe.datetime.add_months(frappe.datetime.get_today(), -3),
    },
    {
      fieldname: "to_date",
      label: __("SO Date To"),
      fieldtype: "Date",
      default: frappe.datetime.get_today(),
    },
  ],

  onload(report) {
    // Wire Select filters to auto-refresh
    setTimeout(() => {
      const f = report.get_filter("status");
      if (f && f.$input) {
        f.$input.off("change.ccr").on("change.ccr", () => report.refresh());
      }
    }, 400);

    report.page.add_action_item(__("Update PTP"), () => open_ptp_update_dialog(report));
  },

  formatter(value, row, column, data, default_formatter) {
    value = default_formatter(value, row, column, data);

    if (column.fieldname === "custom_snrg_credit_check_status" && data) {
      const s = (data.custom_snrg_credit_check_status || "").toLowerCase();
      if (s === "credit hold") {
        value = `<span style="color:#c0392b;font-weight:600;">🔴 ${data.custom_snrg_credit_check_status}</span>`;
      } else if (s === "pending approval") {
        value = `<span style="color:#e67e22;font-weight:600;">🟠 ${data.custom_snrg_credit_check_status}</span>`;
      } else if (s === "approved") {
        value = `<span style="color:#27ae60;font-weight:600;">🟢 ${data.custom_snrg_credit_check_status}</span>`;
      } else if (s === "expired") {
        value = `<span style="color:#f39c12;font-weight:600;">🟡 ${data.custom_snrg_credit_check_status}</span>`;
      } else if (s === "credit ok") {
        value = `<span style="color:#27ae60;">✅ ${data.custom_snrg_credit_check_status}</span>`;
      }
    }

    if (column.fieldname === "ptp_status" && data) {
      const s = (data.ptp_status || "").toLowerCase();
      if (s === "cleared") {
        value = `<span style="color:#27ae60;font-weight:600;">${value}</span>`;
      } else if (s === "partially cleared") {
        value = `<span style="color:#f39c12;font-weight:600;">${value}</span>`;
      } else if (s === "broken") {
        value = `<span style="color:#c0392b;font-weight:600;">${value}</span>`;
      } else if (s === "superseded") {
        value = `<span style="color:#7f8c8d;font-weight:600;">${value}</span>`;
      } else if (s === "pending") {
        value = `<span style="color:#e67e22;font-weight:600;">${value}</span>`;
      }
    }

    // Highlight overdue amounts in red
    if (column.fieldname === "custom_snrg_overdue_amount_terms" && data) {
      if ((data.custom_snrg_overdue_amount_terms || 0) > 0) {
        value = `<span style="color:#c0392b;font-weight:600;">${value}</span>`;
      }
    }

    if (column.fieldname === "ptp_difference_amount" && data) {
      const difference = Number(data.ptp_difference_amount || 0);
      if (difference > 0) {
        value = `<span style="color:#c0392b;font-weight:600;">${value}</span>`;
      } else if (difference <= 0 && Number(data.ptp_received_amount || 0) > 0) {
        value = `<span style="color:#27ae60;font-weight:600;">${value}</span>`;
      }
    }

    // Highlight expired valid-till
    if (column.fieldname === "custom_snrg_override_valid_till" && data) {
      const vt = data.custom_snrg_override_valid_till;
      if (vt && frappe.datetime.get_diff(vt, frappe.datetime.get_today()) < 0) {
        value = `<span style="color:#e74c3c;text-decoration:line-through;">${value}</span>`;
      }
    }

    // Bold PTP commitment date if it has passed and status is still Pending
    if (column.fieldname === "ptp_commitment_date" && data) {
      const cd = data.ptp_commitment_date;
      if (cd && frappe.datetime.get_diff(cd, frappe.datetime.get_today()) < 0) {
        value = `<span style="color:#e74c3c;font-weight:600;" title="PTP date has passed">${value} ⚠️</span>`;
      }
    }

    return value;
  },
};

function open_ptp_update_dialog(report) {
  const row = get_selected_report_row(report);
  const d = new frappe.ui.Dialog({
    title: __("Update PTP"),
    fields: [
      {
        fieldtype: "Link",
        fieldname: "sales_order",
        label: __("Sales Order"),
        options: "Sales Order",
        reqd: 1,
        default: row && row.name ? row.name : "",
        onchange() {
          load_ptp_references_for_sales_order(d, report, d.get_value("sales_order"));
        },
        get_query: () => ({
          filters: {
            docstatus: 0,
            company: report.get_filter_value("company") || undefined,
          },
        }),
      },
      {
        fieldtype: "HTML",
        fieldname: "sales_order_context",
      },
      {
        fieldtype: "Select",
        fieldname: "ptp_reference",
        label: __("PTP Reference"),
        reqd: 1,
        options: "",
      },
      {
        fieldtype: "Link",
        fieldname: "payment_entry",
        label: __("Payment Entry"),
        options: "Payment Entry",
        reqd: 1,
      },
      {
        fieldtype: "Currency",
        fieldname: "allocated_amount",
        label: __("Allocated Amount"),
        reqd: 1,
      },
      {
        fieldtype: "Small Text",
        fieldname: "remarks",
        label: __("Remarks"),
      },
    ],
    primary_action_label: __("Save"),
    primary_action(values) {
      const ref = d._ptp_ref_map && d._ptp_ref_map[values.ptp_reference];
      if (!values.sales_order) {
        frappe.msgprint({
          title: __("Missing Sales Order"),
          message: __("Select a Sales Order."),
          indicator: "red",
        });
        return;
      }
      if (!ref) {
        frappe.msgprint({
          title: __("Missing PTP"),
          message: __("Select a valid PTP reference."),
          indicator: "red",
        });
        return;
      }

      frappe.call({
        method: "snrg_credit_control.overrides.sales_order.link_payment_entry_from_report",
        args: {
          sales_order: values.sales_order,
          ptp_entry_id: ref.ptp_entry_id,
          payment_entry: values.payment_entry,
          allocated_amount: values.allocated_amount,
          remarks: values.remarks,
        },
        freeze: true,
        freeze_message: __("Updating PTP..."),
        callback: () => {
          d.hide();
          frappe.show_alert({ message: __("PTP updated successfully."), indicator: "green" });
          report.refresh();
        },
        error: (err) => {
          frappe.msgprint({
            title: __("Failed to update PTP"),
            message: (err && err.message) || __("Unknown error"),
            indicator: "red",
          });
        },
      });
    },
  });

  d.show();

  d.get_field("ptp_reference").$input.on("change", () => {
    const ref = d._ptp_ref_map && d._ptp_ref_map[d.get_value("ptp_reference")];
    if (!ref) return;
    d.set_value("allocated_amount", Math.max(0, Number(ref.difference_amount || 0)));
  });

  if (d.get_value("sales_order")) {
    load_ptp_references_for_sales_order(d, report, d.get_value("sales_order"));
  }
}

function load_ptp_references_for_sales_order(d, report, salesOrder) {
  if (!salesOrder) {
    d.set_df_property("ptp_reference", "options", "");
    d.set_value("ptp_reference", "");
    d.get_field("sales_order_context").$wrapper.html("");
    d._ptp_ref_map = {};
    return;
  }

  frappe.call({
    method: "snrg_credit_control.overrides.sales_order.get_ptp_references",
    args: { sales_order: salesOrder },
    callback: ({ message }) => {
      const refs = message || [];
      if (!refs.length) {
        d.set_df_property("ptp_reference", "options", "");
        d.set_value("ptp_reference", "");
        d.get_field("sales_order_context").$wrapper.html(
          `<div style="color:var(--text-muted);padding:4px 0;">${__("This Sales Order does not have any PTP entries.")}</div>`
        );
        d._ptp_ref_map = {};
        return;
      }

      const refMap = {};
      const refOptions = refs.map(ref => {
        const label = `${ref.label} | Remaining ${frappe.format(ref.difference_amount || 0, { fieldtype: "Currency", options: report.get_filter_value("currency") || "INR" })} | ${ref.status}`;
        refMap[label] = ref;
        return label;
      });
      d._ptp_ref_map = refMap;
      d.set_df_property("ptp_reference", "options", ["", ...refOptions].join("\n"));
      d.set_value("ptp_reference", "");
      d.set_value("allocated_amount", "");
      d.get_field("payment_entry").get_query = () => ({
        filters: {
          docstatus: 1,
          party_type: "Customer",
          party: infer_customer_from_report(report, salesOrder),
        },
      });
      d.get_field("sales_order_context").$wrapper.html(
        `<div style="background:var(--control-bg);border:1px solid var(--border-color);border-radius:8px;padding:10px 12px;margin-bottom:8px;">
          <div><strong>${__("Sales Order")}:</strong> ${frappe.utils.escape_html(salesOrder)}</div>
        </div>`
      );
    },
  });
}

function get_selected_report_row(report) {
  const datatableRowmanager = report.datatable && report.datatable.rowmanager;
  if (!datatableRowmanager || !datatableRowmanager.getCheckedRows) {
    return null;
  }

  const checkedRows = datatableRowmanager.getCheckedRows() || [];
  if (!checkedRows.length) {
    return null;
  }

  const index = checkedRows[0];
  return (report.data || [])[index] || null;
}

function infer_customer_from_report(report, salesOrder) {
  const row = (report.data || []).find(entry => entry.name === salesOrder);
  return row && row.customer ? row.customer : undefined;
}

File: snrg_credit_control/snrg_credit_control/report/credit_control_report/credit_control_report.py
"""
Credit Control Report
=====================
Shows all draft/open Sales Orders with credit risk or pending approval.

Filters : Company (reqd), Status, Customer, Date From, Date To
Columns : SO, Date, Customer, Order Amount, Credit Status, Reason,
          Overdue Count, Overdue Amount, Request Time,
          Latest PTP (by / payment date / committed amount / mode),
          Approval Status, Approved Cap, Valid Till, Approver
"""

import frappe
from frappe.utils import getdate, today


def execute(filters=None):
    filters = filters or {}
    validate_filters(filters)
    columns = get_columns()
    data    = get_data(filters)
    return columns, data


# ---------------------------------------------------------------------------
# Columns
# ---------------------------------------------------------------------------

def get_columns():
    return [
        {
            "label": "Sales Order",
            "fieldname": "name",
            "fieldtype": "Link",
            "options": "Sales Order",
            "width": 150,
        },
        {
            "label": "SO Date",
            "fieldname": "transaction_date",
            "fieldtype": "Date",
            "width": 100,
        },
        {
            "label": "Customer",
            "fieldname": "customer",
            "fieldtype": "Link",
            "options": "Customer",
            "width": 120,
        },
        {
            "label": "Customer Name",
            "fieldname": "customer_name",
            "fieldtype": "Data",
            "width": 160,
        },
        {
            "label": "Order Amount",
            "fieldname": "grand_total",
            "fieldtype": "Currency",
            "options": "currency",
            "width": 120,
        },
        {
            "label": "Credit Status",
            "fieldname": "custom_snrg_credit_check_status",
            "fieldtype": "Data",
            "width": 120,
        },
        {
            "label": "Reason",
            "fieldname": "custom_snrg_credit_check_reason_code",
            "fieldtype": "Data",
            "width": 110,
        },
        {
            "label": "Overdue #",
            "fieldname": "custom_snrg_overdue_count_terms",
            "fieldtype": "Int",
            "width": 80,
        },
        {
            "label": "Overdue Amount",
            "fieldname": "custom_snrg_overdue_amount_terms",
            "fieldtype": "Currency",
            "options": "currency",
            "width": 130,
        },
        {
            "label": "Request Time",
            "fieldname": "custom_snrg_request_time",
            "fieldtype": "Datetime",
            "width": 140,
        },
        {
            "label": "PTP By",
            "fieldname": "ptp_by_name",
            "fieldtype": "Data",
            "width": 130,
        },
        {
            "label": "PTP Payment Date",
            "fieldname": "ptp_commitment_date",
            "fieldtype": "Date",
            "width": 120,
        },
        {
            "label": "PTP Amount",
            "fieldname": "ptp_committed_amount",
            "fieldtype": "Currency",
            "options": "currency",
            "width": 120,
        },
        {
            "label": "PTP Received",
            "fieldname": "ptp_received_amount",
            "fieldtype": "Currency",
            "options": "currency",
            "width": 120,
        },
        {
            "label": "PTP Difference",
            "fieldname": "ptp_difference_amount",
            "fieldtype": "Currency",
            "options": "currency",
            "width": 130,
        },
        {
            "label": "PTP Status",
            "fieldname": "ptp_status",
            "fieldtype": "Data",
            "width": 130,
        },
        {
            "label": "PTP Mode",
            "fieldname": "ptp_payment_mode",
            "fieldtype": "Data",
            "width": 90,
        },
        {
            "label": "Payment Entries",
            "fieldname": "ptp_payment_entries",
            "fieldtype": "Data",
            "width": 180,
        },
        {
            "label": "Approval Status",
            "fieldname": "custom_credit_approval_status",
            "fieldtype": "Data",
            "width": 110,
        },
        {
            "label": "Approved Cap",
            "fieldname": "custom_snrg_override_cap_amount",
            "fieldtype": "Currency",
            "options": "currency",
            "width": 120,
        },
        {
            "label": "Valid Till",
            "fieldname": "custom_snrg_override_valid_till",
            "fieldtype": "Date",
            "width": 100,
        },
        {
            "label": "Approver",
            "fieldname": "custom_snrg_approver",
            "fieldtype": "Link",
            "options": "User",
            "width": 130,
        },
    ]


# ---------------------------------------------------------------------------
# Data
# ---------------------------------------------------------------------------

def get_data(filters):
    conditions, values = build_conditions(filters)

    so_list = frappe.db.sql(
        f"""
        SELECT
            so.name,
            so.transaction_date,
            so.customer,
            so.customer_name,
            so.grand_total,
            so.rounded_total,
            so.currency,
            so.custom_snrg_credit_check_status,
            so.custom_snrg_credit_check_reason_code,
            so.custom_snrg_overdue_count_terms,
            so.custom_snrg_overdue_amount_terms,
            so.custom_snrg_request_time,
            so.custom_credit_approval_status,
            so.custom_snrg_override_cap_amount,
            so.custom_snrg_override_valid_till,
            so.custom_snrg_approver
        FROM `tabSales Order` so
        WHERE so.docstatus = 0
          {conditions}
        ORDER BY
            FIELD(so.custom_snrg_credit_check_status, 'Credit Hold', 'Not Run', 'Credit OK'),
            so.custom_snrg_request_time DESC,
            so.transaction_date DESC
        """,
        values,
        as_dict=True,
    )

    if not so_list:
        return []

    # Fetch latest PTP entry per SO in one query
    so_names = [r.name for r in so_list]
    placeholders = ", ".join(["%s"] * len(so_names))
    ptp_rows = frappe.db.sql(
        f"""
        SELECT
            p.parent,
            p.name,
            p.ptp_by_name,
            p.commitment_date,
            p.committed_amount,
            p.received_amount,
            p.difference_amount,
            p.status,
            p.linked_payment_entries,
            p.payment_mode,
            p.modified
        FROM `tabCredit PTP Entry` p
        INNER JOIN (
            SELECT parent, MAX(modified) AS latest
            FROM `tabCredit PTP Entry`
            WHERE parent IN ({placeholders})
            GROUP BY parent
        ) latest ON latest.parent = p.parent AND latest.latest = p.modified
        """,
        so_names,
        as_dict=True,
    )
    ptp_map = {r.parent: r for r in ptp_rows}

    today_date = getdate(today())
    data = []
    for so in so_list:
        ptp = ptp_map.get(so.name) or {}
        approval_status = (so.custom_credit_approval_status or "").strip()
        vt = so.custom_snrg_override_valid_till

        # Compute display status (for JS colour coding)
        if so.custom_snrg_credit_check_status == "Credit Hold":
            if approval_status.lower() == "approved":
                if vt and getdate(vt) >= today_date:
                    display_status = "Approved"
                else:
                    display_status = "Expired"
            elif so.custom_snrg_request_time:
                display_status = "Pending Approval"
            else:
                display_status = "Credit Hold"
        else:
            display_status = so.custom_snrg_credit_check_status or "Not Run"

        data.append(
            {
                "name":                               so.name,
                "transaction_date":                   so.transaction_date,
                "customer":                           so.customer,
                "customer_name":                      so.customer_name,
                "grand_total":                        so.grand_total or so.rounded_total or 0,
                "currency":                           so.currency,
                "custom_snrg_credit_check_status":    display_status,
                "custom_snrg_credit_check_reason_code": so.custom_snrg_credit_check_reason_code or "",
                "custom_snrg_overdue_count_terms":    so.custom_snrg_overdue_count_terms or 0,
                "custom_snrg_overdue_amount_terms":   so.custom_snrg_overdue_amount_terms or 0,
                "custom_snrg_request_time":           so.custom_snrg_request_time,
                "ptp_by_name":                        ptp.get("ptp_by_name") or "",
                "ptp_commitment_date":                ptp.get("commitment_date"),
                "ptp_committed_amount":               ptp.get("committed_amount") or 0,
                "ptp_received_amount":                ptp.get("received_amount") or 0,
                "ptp_difference_amount":              ptp.get("difference_amount") or 0,
                "ptp_status":                         ptp.get("status") or "",
                "ptp_payment_mode":                   ptp.get("payment_mode") or "",
                "ptp_payment_entries":                ptp.get("linked_payment_entries") or "",
                "custom_credit_approval_status":      approval_status,
                "custom_snrg_override_cap_amount":    so.custom_snrg_override_cap_amount or 0,
                "custom_snrg_override_valid_till":    vt,
                "custom_snrg_approver":               so.custom_snrg_approver or "",
            }
        )
    return data


# ---------------------------------------------------------------------------
# Filter helpers
# ---------------------------------------------------------------------------

def validate_filters(filters):
    if not filters.get("company"):
        frappe.throw("Please select a Company.")


def build_conditions(filters):
    conditions = "AND so.company = %(company)s"
    values = {"company": filters["company"]}

    if filters.get("customer"):
        conditions += " AND so.customer = %(customer)s"
        values["customer"] = filters["customer"]

    if filters.get("from_date"):
        conditions += " AND so.transaction_date >= %(from_date)s"
        values["from_date"] = filters["from_date"]

    if filters.get("to_date"):
        conditions += " AND so.transaction_date <= %(to_date)s"
        values["to_date"] = filters["to_date"]

    status_filter = filters.get("status")
    if status_filter and status_filter != "All":
        if status_filter == "Credit Hold":
            conditions += (
                " AND so.custom_snrg_credit_check_status = 'Credit Hold'"
                " AND (so.custom_credit_approval_status IS NULL"
                "       OR so.custom_credit_approval_status NOT IN ('Approved', 'Pending'))"
            )
        elif status_filter == "Pending Approval":
            conditions += (
                " AND so.custom_snrg_credit_check_status = 'Credit Hold'"
                " AND so.custom_snrg_request_time IS NOT NULL"
                " AND (so.custom_credit_approval_status IS NULL"
                "       OR so.custom_credit_approval_status != 'Approved')"
            )
        elif status_filter == "Approved":
            conditions += " AND so.custom_credit_approval_status = 'Approved'"
        elif status_filter == "Expired":
            conditions += (
                " AND so.custom_credit_approval_status = 'Approved'"
                " AND so.custom_snrg_override_valid_till < CURDATE()"
            )

    return conditions, values
